networkx/algorithms/approximation/density.py

```diff
@@ -0,0 +1,166 @@
+
+            loads[u] += G_iter.degree[u]  # Add to load of u
+
+            # Subtract one edge from all of u's neighbors, and update their weighted degree
+            for v in G_iter.neighbors(u):
+                weighted_degrees[v] -= 1
+                heap.insert(v, weighted_degrees[v])
+
+            # Delete u
+            G_iter.remove_node(u)
+
+    return (best_density, best_subgraph)
+
+
+ALGORITHMS = {"greedy++": _greedy_plus_plus}
+
+
+@nx.utils.not_implemented_for("directed")
+@nx.utils.not_implemented_for("multigraph")
+@nx._dispatchable
+def densest_subgraph(G, iterations=1, *, method="greedy++"):
+    r"""Returns an approximate densest subgraph for a graph `G`.
+
+    This function runs an iterative algorithm to find
+    the densest subgraph, and returns both the density and the subgraph.
+    For a discussion on the notion of density used and the different
+    algorithms available on networkx, please see the Notes section below.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        Undirected graph.
+
+    iterations : int, optional (default=1)
+        Number of iterations to use for the iterative algorithm.
+        Can be specified positionally or as a keyword argument.
+
+    method : string, optional (default='greedy++')
+        The algorithm to use to approximate the densest subgraph.
+        Supported options: 'greedy++'.
+        Must be specified as a keyword argument. Other inputs produce a ValueError.
+
+    Returns
+    -------
+    d : float
+        The density of the approximate subgraph found.
+
+    S : set
+        The subset of nodes defining the approximate densest subgraph.
+
+    Examples
+    --------
+    >>> G = nx.star_graph(4)
+    >>> nx.approximation.densest_subgraph(G, iterations=1)
+    (0.8, {0, 1, 2, 3, 4})
+
+    Notes
+    -----
+    For a subset of the nodes of $G$, $S \subseteq V(G)$, define
+    $E(S) = \{ (u,v) : (u,v)\in E(G), u\in S, v\in S \}$ as the set of
+    edges with both endpoints in $S$. The density of $S$ is defined as
+    $E(S)/|S|$, the ratio between the edges in the subgraph $G[S]$ and
+    the number of nodes in that subgraph. The densest subgraph problem
+    asks to find the subgraph $S \subseteq V(G)$ with maximum density.
+
+    The densest subgraph problem is polynomial time solvable using
+    maximum flow, commonly refered to as Goldberg's algorithm.
+    However, the algorithm is quite involved. It first
+    binary searches on the optimal density, $d^\ast$. For a guess of
+    the density $d$, it sets up a flow network $G'$ with size O(m). The
+    maximum flow solution either informs the algorithm that no subgraph
+    with density $d$ exists, or it returns (implicitly) a subgraph with density
+    $d$. However, this is inherently bottlenecked by the maximum flow algorithm.
+    For example, authors from [2]_ noted that Goldberg’s algorithm
+    failed on many large graphs even though they used a highly optimized
+    maximum flow library.
```

Comment:
- Maybe  `returns (implicitly)` -> `provides` ?? WDYT?
- Do the "authors of [2] noted that ..." in a publication other than [2]? If the note is in [2], then just refer to [2].
- What does "failed" mean here? Even if it takes a long time, it doesn't fail, right? Or if memory is the problem that is a different issue -- but still not really a failure so much as a need for better hardware. I suggest the word "feasible" below, but that might have technical meanings I'm not aware of. So what is a better word?

```suggestion
    with density $d$ exists, or it provides a subgraph with density at least $d$.
    This process is inherently bottlenecked by the maximum flow algorithm.
    For example, [2]_ notes that Goldberg’s algorithm was not feasible
    on many large graphs even though they used a highly optimized
    maximum flow library.
```